docs: backport inflight-window-and-message-queue.md (#2845)
emqtt/emq-docs-cn: en_US/design/inflight-window-and-message-queue.md

# In-flight and Message Queue

## Introduction

To improve message throughput and reduce the impact of network fluctuations, EMQX allows multiple unacknowledged QoS 1 and QoS 2 messages in-flight at the same time. These sent but unconfirmed messages will be stored in the Inflight Window until acknowledgment is complete.

When the number of in-flight messages exceeds the limit, that is, the length limit of Inflight Window is reached(see `max_inflight`), EMQX will no longer send subsequent messages, but will store these messages in the Message Queue. Once a message is acknowledged in the Inflight Window, the message in the Message Queue will be sent in first-in, first-out order and stored in the Inflight Window.

If the number of in-flight QoS 1, 2 messages reaches the maximum limit of the Inflight Window (see `max_inflight`), the newly arrived messages will not be forwarded immediately, but will be temporarily stored in the Message Queue. 

> Only when previous messages are confirmed and removed from the Inflight Window, the messages in the Message Queue will be sent in a FIFO order and added to the Inflight Window. QoS 0 messages are not affected by this, they are always forwarded immediately.

Reviewer: ```suggestion
Messages in the Message Queue are sent in FIFO order and added to the Inflight Window only after previous messages have been confirmed and removed. QoS 0 messages, however, are not affected by this process and are always forwarded immediately.
```

However, EMQX does not necessarily grant the `Receive Maximum` value requested in the CONNECT packet. Instead, the `Receive Maximum` granted in the CONNACK packet is capped by the `mqtt.max_inflight` configuration.

## Configuration Items

| Configuration Items    | Type    | Optional Value  | Default Value | Description                                                                       |
| -----------------------| ------- | --------------- | ------------------------------------------------------------------------------------------------- |
| mqtt.max_inflight      | integer | (0, 65536)      | 32            | Inflight Window length limit, 0 means no limit                                    |
| mqtt.max_mqueue_len    | integer | [0, ∞)          | 1000          | Message Queue length limit, 0 means no limit                                      |

> | mqtt.mqueue_store_qos0 | enum    | `true`, `false` | true          | Whether EMQX store QoS 0 messages to the Message Queue when the client is offline |

Reviewer: ```suggestion
| mqtt.mqueue_store_qos0 | enum    | `true`, `false` | true          | Whether EMQX stores QoS 0 messages to the Message Queue when the client is offline |
```
